### PR TITLE
- StorageAllocation optimisation - don't need Blobbers

### LINKED ISSUE #1114: Optimise StorageAllocation
Try to understand the use of storage allocation and optimise it's corresponding smartcontracts.

```
// StorageAllocation request and entity.
type StorageAllocation struct {
	// ID is unique allocation ID that is equal to hash of transaction with
	// which the allocation has created.
	ID string `json:"id"`
	// Tx keeps hash with which the allocation has created or updated.
	Tx string `json:"tx"`

	DataShards        int                           `json:"data_shards"`
	ParityShards      int                           `json:"parity_shards"`
	Size              int64                         `json:"size"`
	Expiration        common.Timestamp              `json:"expiration_date"`
	Blobbers          []*StorageNode                `json:"blobbers"`
	Owner             string                        `json:"owner_id"`
	OwnerPublicKey    string                        `json:"owner_public_key"`
	Stats             *StorageAllocationStats       `json:"stats"`
	DiverseBlobbers   bool                          `json:"diverse_blobbers"`
	PreferredBlobbers []string                      `json:"preferred_blobbers"`
	BlobberDetails    []*BlobberAllocation          `json:"blobber_details"`
	BlobberMap        map[string]*BlobberAllocation `json:"-" msg:"-"`
	IsImmutable       bool                          `json:"is_immutable"`

	// Requested ranges.
	ReadPriceRange             PriceRange    `json:"read_price_range"`
	WritePriceRange            PriceRange    `json:"write_price_range"`
	MaxChallengeCompletionTime time.Duration `json:"max_challenge_completion_time"`

	//AllocationPools allocationPools `json:"allocation_pools"`
	WritePoolOwners []string `json:"write_pool_owners"`

	// ChallengeCompletionTime is max challenge completion time of
	// all blobbers of the allocation.
	ChallengeCompletionTime time.Duration `json:"challenge_completion_time"`
	// StartTime is time when the allocation has been created. We will
	// use it to check blobber's MaxOfferTime extending the allocation.
	StartTime common.Timestamp `json:"start_time"`
	// Finalized is true where allocation has been finalized.
	Finalized bool `json:"finalized,omitempty"`
	// Canceled set to true where allocation finalized by cancel_allocation
	// transaction.
	Canceled bool `json:"canceled,omitempty"`
	// UsedSize used to calculate blobber reward ratio.
	UsedSize int64 `json:"-" msg:"-"`

	// MovedToChallenge is number of tokens moved to challenge pool.
	MovedToChallenge state.Balance `json:"moved_to_challenge,omitempty"`
	// MovedBack is number of tokens moved from challenge pool to
	// related write pool (the Back) if a data has deleted.
	MovedBack state.Balance `json:"moved_back,omitempty"`
	// MovedToValidators is total number of tokens moved to validators
	// of the allocation.
	MovedToValidators state.Balance `json:"moved_to_validators,omitempty"`

	// TimeUnit configured in Storage SC when the allocation created. It can't
	// be changed for this allocation anymore. Even using expire allocation.
	TimeUnit time.Duration `json:"time_unit"`

	Curators []string `json:"curators"`
}
```

### PR DESCRIPTION
**Pull request by ssardana96**
## Fixes
- Removed use of Blobbers in 0chain sc
- Blobbers would only be required for /allocation and /allocations client API
- also lot of gosdk depends on /allocation API - which will lead to huge changes if we remove Blobbers from there.
- Area of impact - MPT data size (which will help in sc execution)

## Tests 
Tasks to complete before merging PR:
- [x]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [x]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [x]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber: https://github.com/0chain/blobber/pull/623
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | feat/generate_challenge  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | sa_optimise  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/storagesc/allocation.go [hunk 1]
```diff
@@ -365,7 +365,6 @@ func (sc *StorageSmartContract) newAllocationRequestInternal(
 		return allocatedBlobbers[i].ID < allocatedBlobbers[j].ID
 	})
 
-	sa.Blobbers = allocatedBlobbers
```
Comment (Sriep) [inline code/go/0chain.net/smartcontract/storagesc/allocation.go:361]: allocatedBlobbers does not seem to be used anywhere, so we should be able to remove it.

**Review by Sriep — DISMISSED** [review]

**Review by peterlimg — DISMISSED** [review]
lgtm

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | sa_optimise  |
| blobber  | sa_optimise  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [success] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | sa_optimise  |
| blobber  | sa_optimise  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | staging  |
| 0proxy  | staging  |
| 0search  | staging  |

**Review by peterlimg — DISMISSED** [review]
lgtm

**Review by peterlimg — APPROVED** [review]
lgtm

File: code/go/0chain.net/smartcontract/storagesc/models.go [hunk 2]
```diff
@@ -836,24 +838,12 @@ func (sa *StorageAllocation) removeBlobber(
...
-	if !found {
-		return nil, fmt.Errorf("cannot find blobber %s in allocation", remove.BlobberID)
-	}
 	var removedBlobber *StorageNode
 	found = false
 	for i, d := range blobbers {
 		if d.ID == removeId {
 			removedBlobber = blobbers[i]
-			blobbers[i] = blobbers[len(sa.Blobbers)-1]
+			blobbers[i] = blobbers[len(sa.BlobberDetails)-1]
```
Comment (peterlimg) [inline code/go/0chain.net/smartcontract/storagesc/models.go:846]: Incorrect assignment, the previous code was correct.

Comment (ssardana96) [reply code/go/0chain.net/smartcontract/storagesc/models.go:846]: why is this wrong @peterlimg ?
If you track back this function call, it is reading from blobberDetails and removing from there..

Comment (peterlimg) [reply code/go/0chain.net/smartcontract/storagesc/models.go:846]: Ah, yes, but this is true only when blobbers and sa.BlobberDetails has the same item numbers, even they are the same, we'd better use the `blobbers` itself, like `blobbers[i] = blobbers[len(blobbers) - 1]`. What we want to do is replace the `i` item with the last item, so shouldn't bother other variables.

Comment (ssardana96) [reply code/go/0chain.net/smartcontract/storagesc/models.go:846]: ahh got it!
yes makes sense..

I'll change this in any of my next PR

**Review by peterlimg — COMMENTED** [review]
I found this PR was intended to remove the `[]Blobbers`, but later added it back, so the only change is changed the AllocationStatsHandler to get blobber from event db, am I right?

**Comment by ssardana96** [comment]
> I found this PR was intended to remove the `[]Blobbers`, but later added it back, so the only change is changed the AllocationStatsHandler to get blobber from event db, am I right?

yes @peterlimg 
ideally blobber data should not be stored in MPT.
Just that our API expects blobbers so added it back and getting from eventsdb.

**Comment by peterlimg** [comment]
@ssardana08 if the `[]Blobbers` is only used for API, then we should remove it from the data struct and create an new data struct for API use only. We should not keep unused fields in MPT.

**Pull request merged**